db/schema.rb

```diff
@@ -122,14 +123,17 @@
     t.string "video_source_url"
     t.string "video_state"
     t.string "video_thumbnail_url"
+    t.index "digest(body_markdown, 'sha512'::text), user_id, title", name: "index_articles_on_digest_body_markdown_and_user_id_and_title", unique: true
```

Comment:
Even though I loved the exploration in your post, I think we may want to question whether `body_markdown` should actually be "unique" from a usability perspective.

We've had this in, but I'm not sure it's the ideal behavior in all cases. I think getting the logic into the database is the point of no return vs the rails validators which are more flexible.

I think the biggest reason this was originally added to the validators was to deal with the mishap of situations where folks could accidentally submit the same new post twice depending on how they hit refresh or whatever in the browser at the wrong time... Which could still be avoided.

But I could see a legit case for a "repeating post" or something along those lines where you don't want the `body_markdown` to be entirely unique.

Reply:
The main advantage of having this in the DB is that there's still not a 100% guarantee that doing so in Rails won't lead to duplicate posts. Mainly because of concurrency. The chances are very slim, but double posting is not completely unheard of, which is why we have the uniqueness check anyway. I don't think it's completely a point of no-return as we can easily remove it with an `alter table`.

Rails validators would stay in, the whole idea is to have both. But a Rails validator is going to be less performing if it's not backed by a unique index. While I was exploring this I ended up in Rails source code and what Rails basically does is to do a `SELECT` with all the scoped fields (`body_markdown`, `user_id`, `title` in this instance), see if it returns something and if it does, they trigger the validator, if it doesn't they let the operation go through.

This is where the unique index inserts itself:

- it quickens that select by using the index instead of walking through all rows
- it avoid the improbable but still impossible case of an insert happening at the same time (the double posting)

We likely found thousands of duplicates (most likely related to the RSS importer, not probably due to doubleposting) in the database which is probably why a validator alone is not sufficient.

I still have to investigate why we have so many duplicates (how they got in).